DOCUMENTATION.md

```diff
@@ -1370,529 +1395,202 @@ The following are example configurations.
+```
 
-## 5.8. Enabling Hazelcast Cloud Discovery
+## 6.1. Configuring Client Connection Retry
 
-The purpose of Hazelcast Cloud Discovery is to provide the clients to use IP addresses provided by `hazelcast orchestrator`. To enable Hazelcast Cloud Discovery, specify a token for the `discoveryToken` field and set the `enabled` field to `true`.
+When client is disconnected from the cluster, it searches for new connections to reconnect.
+You can configure the frequency of the reconnection attempts and client shutdown behavior using
+`ConnectionRetryConfig` (programmatical approach) or `connectionRetry` element (declarative approach).
 
-The following are example configurations.
+Below are the example configurations for each.
 
 **Declarative Configuration:**
 
 ```json
 {
- "group": {
-        "name": "hazel",
-        "password": "cast"
-    },
-
-    "network": {
-        "hazelcastCloud": {
-            "discoveryToken": "EXAMPLE_TOKEN",
-            "enabled": true
+    "connectionStrategy": {
+        "asyncStart": false,
+        "reconnectMode": "ON",
+        "connectionRetry": {
+            "initialBackoffMillis": 1000,
+            "maxBackoffMillis": 60000,
+            "multiplier": 2,
+            "clusterConnectTimeoutMillis": 50000,
+            "jitter": 0.2
         }
     }
 }
-
 ```
 
 **Programmatic Configuration:**
 
 ```javascript
 var clientConfig = new Config.ClientConfig();
-clientConfig.groupConfig.name = 'hazel';
-clientConfig.groupConfig.password = 'cast';
+var connectionRetryConfig = new Config.ConnectionRetryConfig();
+connectionRetryConfig.initialBackoffMillis = 1000;
+connectionRetryConfig.maxBackoffMillis = 60000;
+connectionRetryConfig.multiplier = 2;
+connectionRetryConfig.clusterConnectTimeoutMillis = 50000;
+connectionRetryConfig.jitter = 0.2;
 
-clientConfig.networkConfig.cloudConfig.enabled = true;
-clientConfig.networkConfig.cloudConfig.discoveryToken = 'EXAMPLE_TOKEN';
+clientConfig.connectionStrategyConfig.connectionRetryConfig = connectionRetryConfig;
 ```
 
-To be able to connect to the provided IP addresses, you should use secure TLS/SSL connection between the client and members. Therefore, you should set an SSL configuration as described in the previous section.
+The following are configuration element descriptions:
 
-# 6. Securing Client Connection
+* `initialBackoffMillis`: Specifies how long to wait (backoff), in milliseconds, after the first failure before retrying. Its default value is 1000 ms.
+* `maxBackoffMillis`: Specifies the upper limit for the backoff in milliseconds. Its default value is 30000 ms.
+* `multiplier`: Factor to multiply the backoff after a failed retry. Its default value is 1.
+* `clusterConnectTimeoutMillis`: Timeout value in milliseconds for the client to give up to connect to the current cluster Its default value is 20000.
+* `jitter`: Specifies by how much to randomize backoffs. Its default value is 0.
```

Comment:
nit: I guess that `jitter` should be in `[0.0, 1.0]` range, right? We should mention it here.

Reply:
Looking at the Java side and implementation, I think there is no range limit for jitter. It is used as a multiplicand while determining the actual sleep time. It is probably not sane to set it more than 1.0 but I didn't see a limiting factor for it.

Reply:
If `jitter > 1.0`, then we may get a negative time period for the sleep interval which makes no sense. So, we should describe the range, at least in docs (some config validation would be even better). As for the Java client, it makes sense to submit a PR in order to improve the documentation and config validation. WDYT?

Reply:
Yes you are right, it should be in that range. Also, there is a bug in the implementation of backoff time with jitter in the java side (and in this PR). I will send a fix for both of them.

Reply:
By the way, I have a general question. Since there is a range limit for this property, what should we do for the programmatic configuration of these kinds of elements ?. Right now, we are not using setters for most of the config elements and user can set them to any value he wants which may result in failures. Should we switch all of them to setters ?

Reply:
I think it's better to leave them as properties. ES6 allows to define setters (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/set), but I don't think we need them here.

Ideally, the config should be a plain JS object and we should validate it when `newHazelcastClient()` function is called. Unfortunately, that's not what we have right now.

As for this particular field (`jitter`) and other similar ones, let's omit validation for it for now.